### PR TITLE
Allow to deserialize an empty string to null

### PR DESCRIPTION
**Pull request by dstepanov**
Not sure if this is the best solution. Maybe it should be configurable?

Fixes https://github.com/micronaut-projects/micronaut-serialization/issues/619

### CODE REVIEW

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java [hunk 1]
```diff
@@ -292,6 +292,17 @@ private Boolean decodeBooleanSlow() throws IOException {
                 return null;
             }
             case START_OBJECT, END_OBJECT, END_ARRAY, FIELD_NAME -> throw unexpectedToken(JsonToken.VALUE_TRUE, t);
+            case VALUE_STRING -> {
+                String string = parser.getText();
+                if (string.isEmpty()) {
```
Comment (sdelamo) [inline serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:297]: Should we use?

```suggestion
                if (StringUtils.isEmpty(string)) {
```

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java [hunk 2]
```diff
@@ -312,6 +323,17 @@ public byte decodeByte() throws IOException {
     public Byte decodeByteNullable() throws IOException {
         JsonToken t = nextToken();
         switch (t) {
+            case VALUE_STRING -> {
+                String string = parser.getText();
+                if (string.isEmpty()) {
```
Comment (sdelamo) [inline serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:328]:
```suggestion
                if (StringUtils.isEmpty(string)) {
```

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java [hunk 3]
```diff
@@ -353,6 +375,17 @@ public short decodeShort() throws IOException {
     public Short decodeShortNullable() throws IOException {
         JsonToken t = nextToken();
         switch (t) {
+            case VALUE_STRING -> {
+                String string = parser.getText();
+                if (string.isEmpty()) {
```
Comment (sdelamo) [inline serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:380]:
```suggestion
                if (StringUtils.isEmpty(string)) {
```

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java [hunk 4]
```diff
@@ -407,6 +440,9 @@ public Character decodeCharNullable() throws IOException {
             }
             case VALUE_STRING -> {
                 String string = parser.getText();
+                if (string.isEmpty()) {
```
Comment (sdelamo) [inline serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:443]:
```suggestion
                if (StringUtils.isEmpty(string)) {
```

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java [hunk 5]
```diff
@@ -460,6 +496,9 @@ private Integer decodeIntSlow() throws IOException {
...
```
Comment (sdelamo) [inline serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:499]:
```suggestion
                if (StringUtils.isEmpty(string)) {
```

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java [hunk 6]
```diff
@@ -528,13 +567,14 @@ private Long decodeLongSlow() throws IOException {
             }
             case VALUE_STRING -> {
                 String string = parser.getText();
-                long value;
+                if (string.isEmpty()) {
```
Comment (sdelamo) [inline serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:570]:
```suggestion
                if (StringUtils.isEmpty(string)) {
```

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java [hunk 7]
```diff
@@ -579,13 +619,14 @@ public Float decodeFloatNullable() throws IOException {
         switch (t) {
             case VALUE_STRING -> {
                 String string = parser.getText();
-                float value;
+                if (string.isEmpty()) {
```
Comment (sdelamo) [inline serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:622]:
```suggestion
                if (StringUtils.isEmpty(string)) {
```

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java [hunk 8]
```diff
@@ -641,6 +682,9 @@ private Double decodeDoubleSlow(JsonToken t) throws IOException {
             }
             case VALUE_STRING -> {
                 String string = parser.getText();
+                if (string.isEmpty()) {
```
Comment (sdelamo) [inline serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:685]:
```suggestion
                if (StringUtils.isEmpty(string)) {
```

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java [hunk 9]
```diff
@@ -691,6 +735,9 @@ public BigInteger decodeBigIntegerNullable() throws IOException {
         switch (t) {
             case VALUE_STRING -> {
                 String string = parser.getText();
+                if (string.isEmpty()) {
```
Comment (sdelamo) [inline serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:738]:
```suggestion
                if (StringUtils.isEmpty(string)) {
```

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java [hunk 10]
```diff
@@ -741,6 +788,9 @@ public BigDecimal decodeBigDecimalNullable() throws IOException {
...
```
Comment (sdelamo) [inline serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:791]:
```suggestion
                if (StringUtils.isEmpty(string)) {
```

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java [hunk 11]
```diff
@@ -793,7 +843,8 @@ private Number decodeNumber() throws IOException {
 
     @Override
     public boolean decodeNull() throws IOException {
-        if (peekToken() == JsonToken.VALUE_NULL) {
+        JsonToken jsonToken = peekToken();
+        if (jsonToken == JsonToken.VALUE_NULL || jsonToken == JsonToken.VALUE_STRING && parser.getText().isBlank()) {
```
Comment (sdelamo) [inline serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:847]:
```suggestion
        if (jsonToken == JsonToken.VALUE_NULL || jsonToken == JsonToken.VALUE_STRING && StringUtils.isEmpty(parser.getText())) {
```

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java (hunk 1 above)
Comment (yawkat) [reply serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:297]: why?

**Comment by yawkat** [comment]
why on earth does JacksonDecoder matter for forms?

**Comment by dstepanov** [comment]
To convert forms into a bean, we convert a map into the JSON node model in `JsonBeanPropertyBinder` and deserialize it into a bean. And we get all kind of problems like this and https://github.com/micronaut-projects/micronaut-serialization/issues/616

**Review by graemerocher — APPROVED** [review]

**Pull request merged**